source/winAPI/messageWindow.py

```diff
@@ -4,19 +4,152 @@
+@type wParam: int
+@param lParam: Additional message information.
+@type lParam: int
+"""
+
+
+class WindowMessage(IntEnum):
+	DISPLAY_CHANGE = 0x7e
 	"""
+	WM_DISPLAYCHANGE
+
+	https://docs.microsoft.com/en-us/windows/win32/gdi/wm-displaychange
+	"""
+	POWER_BROADCAST = 0x0218
+	"""
+	WM_POWERBROADCAST
+
+	An application should process this event by calling the GetSystemPowerStatus function
+	to retrieve the current power status of the computer.
+	In particular, the application should check the
+	ACLineStatus, BatteryFlag, BatteryLifeTime, and BatteryLifePercent members
+	of the SYSTEM_POWER_STATUS structure for any changes.
+
+	This event can occur when battery life drops to less than 5 minutes,
+	or when the percentage of battery life drops below 10 percent,
+	or if the battery life changes by 3 percent.
+
+	https://docs.microsoft.com/en-us/windows/win32/power/pbt-apmpowerstatuschange
+	"""
+
+	WTS_SESSION_CHANGE = 0x02B1
+	"""
+	WM_WTSSESSION_CHANGE
+
 	Windows Message for when a Session State Changes.
-	Receiving these messages is registered by registerSessionNotification.
+	Receiving these messages is registered by sessionTracking.register.
 	handleSessionChange handles these messages.
+
+	https://docs.microsoft.com/en-us/windows/win32/api/wtsapi32/nf-wtsapi32-wtsregistersessionnotification
+	"""
+
+
+class _MessageWindow(windowUtils.CustomWindow):
+	className = "wxWindowClassNR"
+	"""
+	#3763: In wxPython 3, the class name of frame windows changed from wxWindowClassNR to wxWindowNR.
+	NVDA uses the main frame to check for and quit another instance of NVDA.
+	To remain compatible with older versions of NVDA, create our own wxWindowClassNR.
+	We don't need to do anything else because wx handles WM_QUIT for all windows.
 	"""
+
+	def __init__(self, windowName: Optional[str] = None):
+		super().__init__(windowName)
+		powerTracking.initialize()
+
+		# Call must be paired with a call to sessionTracking.unregister
+		self._isSessionTrackingRegistered = sessionTracking.register(self.handle)
+
+		# Call must be paired with calls to deregister
+		post_windowMessageReceipt.register(_MessageWindow.handleWindowMessage)
+
+	def warnIfSessionTrackingNotRegistered(self) -> None:
+		if self._isSessionTrackingRegistered:
+			return
+		failedToRegisterMsg = _(
+			# Translators: This is a warning to users, shown if NVDA cannot determine if
+			# Windows is locked.
+			"NVDA failed to register session tracking. "
+			"While this instance of NVDA is running, "
+			"your desktop will not be secure when Windows is locked. "
+			"Restart NVDA? "
+		)
+		if wx.YES == gui.messageBox(
+			failedToRegisterMsg,
+			# Translators: This is a warning to users, shown if NVDA cannot determine if
+			# Windows is locked.
+			caption=_("NVDA could not start securely."),
+			style=wx.ICON_ERROR | wx.YES_NO,
+		):
+			import core
+			core.restart()
+
+	def destroy(self):
+		"""
+		NVDA must unregister session tracking before destroying the message window.

```

Comment:
It would be helpful to explain why. But I don't see why using the extension point class is helpful in this case, see my other comment.

Reply:
Sorry, looks like I put this on the wrong line. The comment this was intended for has been removed.